storage/imx6ull-flashnor/flashnor-qspi.c

```diff
@@ -0,0 +1,303 @@
+	return -ETIME;
+}
+
+
+ssize_t flashnor_qspiRead(qspi_dev_t dev, unsigned int addr, void *buff, size_t bufflen)
+{
+	size_t size, len;
+	int err;
+
+	err = mutexLock(flashnor_common.lock);
+	if (err < 0) {
+		return err;
+	}
+
+	for (len = 0; len < bufflen; len += size) {
+		if ((size = bufflen - len) > MAX_READ_LEN)
+			size = MAX_READ_LEN;
+
+		/* TODO XIP */
+		err = _qspi_readBusy(dev, lut_seq_read, addr + len, ((char *)buff) + len, size);
+		if (err < 0) {
+			mutexUnlock(flashnor_common.lock);
+			return err;
+		}
+	}
+
+	err = mutexUnlock(flashnor_common.lock);
+	if (err < 0) {
+		return err;
+	}
+
+	return bufflen;
+}
+
+
+ssize_t flashnor_qspiWrite(qspi_dev_t dev, unsigned int addr, const void *buff, size_t bufflen)
+{
+	size_t size, len;
+	int err;
+
+	err = mutexLock(flashnor_common.lock);
+	if (err < 0) {
+		return err;
+	}
+
+	for (len = 0; len < bufflen; len += size) {
+		err = _flashnor_qspiWriteEnable(dev);
+		if (err < 0) {
+			mutexUnlock(flashnor_common.lock);
+			return err;
+		}
+		size = bufflen - len;
+		if (size > MAX_WRITE_LEN)
+			size = MAX_WRITE_LEN;
+
+		/* Limit write size to the page aligned address (don't wrap around at the page boundary) */
+		if (size > FLASH_PAGE_SIZE - ((addr + len) & (FLASH_PAGE_SIZE - 1))) {
+			size = FLASH_PAGE_SIZE - ((addr + len) & (FLASH_PAGE_SIZE - 1));
+		}
+		/* TODO XIP */
+		err = _qspi_write(dev, lut_seq_write, addr + len, ((const char *)buff) + len, size);
+		if (err < 0) {
+			mutexUnlock(flashnor_common.lock);
+			return err;
+		}
+		err = _flashnor_qspiWaitBusy(dev);
+		if (err < 0) {
+			mutexUnlock(flashnor_common.lock);
+			return err;
+		}
+	}
+
+	err = mutexUnlock(flashnor_common.lock);
+	if (err < 0) {
+		return err;
+	}
+
+	return bufflen;
+}
+
+
+int flashnor_qspiEraseSector(qspi_dev_t dev, unsigned int addr)
+{
+	int err;
+
+	err = mutexLock(flashnor_common.lock);
+	if (err < 0) {
+		return err;
+	}
+	err = _flashnor_qspiWriteEnable(dev);
+	if (err < 0) {
+		mutexUnlock(flashnor_common.lock);
+		return err;
+	}
+	err = _qspi_write(dev, lut_seq_erase, addr, NULL, 0);
+	if (err < 0) {
+		mutexUnlock(flashnor_common.lock);
+		return err;
+	}
+	err = _flashnor_qspiWaitBusy(dev);
+	if (err < 0) {
+		mutexUnlock(flashnor_common.lock);
+		return err;
+	}
+	return mutexUnlock(flashnor_common.lock);
+}
+
+
+static int get_jedec_id(qspi_dev_t dev, uint8_t data[3])
+{
+	int err;
+	static const lut_seq_t seq = { .instrs = { LUT_INSTR(lut_cmd, lut_pad1, CMD_JEDEC), LUT_INSTR(lut_read, lut_pad1, 3), 0 } };
+
+	err = _qspi_setLutSeq(&seq, lut_seq_jedec);
+	if (err < 0) {
+		return err;
+	}
+	return _qspi_readBusy(dev, lut_seq_jedec, 0, data, 3);
+}
+
+
+static int enable_quad_io(qspi_dev_t dev)
+{
+	int err;
+	static const lut_seq_t seq = { .instrs = { LUT_INSTR(lut_cmd, lut_pad1, CMD_QENABLE), 0 } };
+	err = _qspi_setLutSeq(&seq, lut_seq_quad_io);
+	if (err < 0) {
+		return err;
+	}
+	return _qspi_readBusy(dev, lut_seq_quad_io, 0, NULL, 0);
+}
+
+
+static int populate_lut(void)
```

Comment:
Nitpick - function being unlocked, but no `_` prefix used. We have mutex in this module, so it makes sense to mark functions as unlocked.

It is advised to use module prefix for static functions too - you can still cause a bug, you can collide with some external function (obvious example - naming a function `read()`).

Use camelCase
`
Applies to functions above too.